### ISSUE TITLE
Cannot find system headers on macOS

### ISSUE THREAD

**Issue by mohd-akram** [include-what-you-use/include-what-you-use#1308]
I have a simple C file:

```c
#include <stdio.h>
int main(void)
{
	puts("hello world");
	return 0;
}
```

Running `include-what-you-use main.c` nets:

```c
main.c:1:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^~~~~~~~~
```

I'm running version 0.20 with MacPorts Clang 16 on macOS Ventura. I can confirm that Clang can build the file via `clang-mp-16 main.c`. How does it look for headers?

EDIT: It was working in clang because `clang-mp-16` is a shell script that runs the actual clang via `xcrun`. See my comment below on ways to resolve this for `include-what-you-use`.

**Comment by kimgr** [comment]
I recently added a little section on platform tweaks to the readme: https://github.com/include-what-you-use/include-what-you-use#platform-tweaks.

But after publishing that, I realized that the easiest way to use IWYU on Mac is to use the `-isysroot` switch:
```
include-what-you-use -isysroot $(xcrun --sdk macosx --show-sdk-path) main.c
```
(here using Bash substitutions, not sure if that works with zsh, for example).

Setting the sysroot takes care of both the C headers and the C++ headers.

**Comment by mohd-akram** [comment]
I investigated the problem more deeply. There are two issues:

- Clang needs to find system directories. It does this via the `SDKROOT` environment variable. Running via `xcrun` sets `SDKROOT` (try eg. `xcrun env`), so one can run `include-what-you-use` like so:

   ```shell
   xcrun include-what-you-use main.c
   ```

   Alternatively, add this to the shell profile:

   ```shell
   alias include-what-you-use="xcrun include-what-you-use"
   ```

   or:

   ```shell
   export SDKROOT=$(xcrun --show-sdk-path)
   ```

  or `clang` must be built with the `-DDEFAULT_SYSROOT` option.

- Clang needs to find its own include directories. This is solved by installing `include-what-you-use` next to the clang binary so that it can find the `include` directory in the level above. This can potentially be solved in the tool itself by changing the path passed to the `Driver` constructor to be the same as `clang`'s path. I don't know if the clang library exposes that information. I was also looking into using `dladdr` to get the path of the linked clang library and finding it that way.

**Comment by kimgr** [comment]
Thanks, those are nice techniques. See also issue #100, which covers the challenges with the builtin headers.

How did you install IWYU? I've only built from source on Mac, so maybe my experience is atypical.

**Comment by mohd-akram** [comment]
I've built from source as well. I was creating a MacPorts port for it and ran into these issues as well. The two changes I made to make it work out of the box were to move the binary next to clang and create a wrapper script that runs it via `xcrun`, so this is solvable from a packaging perspective. It can be installed now via `sudo port install include-what-you-use`.

Thanks for sharing that thread. I think a simple fix is to have a `IWYU_CLANG_PATH` option that defaults to `${CLANG_INSTALL_PREFIX}/bin/clang` in CMake and pass it as a macro definition onto the program. Then just ifdef `IWYU_CLANG_PATH` when setting up the driver and if set use that as the driver path otherwise fallback to argv0. Then it should theoretically behave *exactly* as clang does (I tried hardcoding the driver path to the location of my clang and it does work to resolve header files). Using `-resource-path` does not work for C++ headers as those are [hardcoded](https://github.com/llvm/llvm-project/blob/llvmorg-16.0.6/clang/lib/Driver/ToolChains/Darwin.cpp#L2425) to be relative to the install location.

**Comment by kimgr** [comment]
> I think a simple fix is to have a IWYU_CLANG_PATH option that defaults to ${CLANG_INSTALL_PREFIX}/bin/clang in CMake and pass it as a macro definition onto the program.

That would work for MacPorts only because we're building on the target system, right? `CLANG_INSTALL_PREFIX` would be provided by an earlier installed Clang CMake config, so it be wired to the actual install location?

I'm thinking it doesn't generalize well to packaging where binaries are built and distributed, because the package building system may have Clang in a different location than the target system.

I've been hoping to come up with some mechanism that supports both scenarios -- I think the compile-on-target use case is actually more unusual. Happy to be proven wrong there, because it would be pretty simple to set up!

For C++ headers, `-isysroot` should work; the function you linked to proceeds to search any provided sysroot for C++ header paths.

**Comment by mohd-akram** [comment]
> That would work for MacPorts only because we're building on the target system, right? CLANG_INSTALL_PREFIX would be provided by an earlier installed Clang CMake config, so it be wired to the actual install location?

True, but that would just be a default value for `IWYU_CLANG_PATH` for convenience. It could also simply have no default value, and in that case the package would be built with the argv0 behavior by default - maintaining the current behavior - while those who know where clang would be can specify it. So essentially, for the two camps, you'd have:

- Install location specified at build time: specify `IWYU_CLANG_PATH` to be relative to already-installed include files
- Install location specified at install time: package include files and install them relative to the binary

> I'm thinking it doesn't generalize well to packaging where binaries are built and distributed, because the package building system may have Clang in a different location than the target system.

Packaging systems in the UNIX world generally depend on things being where they are specified to be (eg. via [`CMAKE_INSTALL_PREFIX`](https://cmake.org/cmake/help/latest/variable/CMAKE_INSTALL_PREFIX.html)) at build time. This is because the prefix tends to be embedded in many areas, such as in the path to linked libraries, configuration file paths, documentation, and so on. This is why if you use a non-default prefix in eg. MacPorts or Homebrew, you have to [build everything from source](https://docs.brew.sh/FAQ#why-should-i-install-homebrew-in-the-default-location) because their pre-built packages have the default prefix embedded all over the place.

**Comment by mohd-akram** [comment]
Hopefully, I'll see if I can get to this, my CMake knowledge isn't the best :)

**Comment by kimgr** [comment]
What with #1637 to make IWYU's relationship to clang builtin headers more configurable, and #1432 to expand on macOS use cases, I'm calling this fixed.
